python/setup.py

```diff
@@ -54,7 +55,42 @@ def _copy_backends(active):
         package_data = [f"{os.path.relpath(p, backend_path)}/*" for p, _, _, in os.walk(backend_path)]
         ret.append(
             Backend(name=backend, package_data=package_data, src_dir=curr_path, backend_dir=backend_path,
-                    install_dir=install_dir))
+                    install_dir=install_dir, is_external=False))
+    return ret
+
+def _copy_external_backends():
+    # TRITON_EXTERNAL_CODEGEN_BACKENDS is a semicolon-separated list of backend names
+    # TRITON_EXTERNAL_CODEGEN_BACKEND_DIRS is a semicolon-separated list of paths to the backends
+    # Note that the paths have to include the provided backend names as a subfolder
+    backends = os.getenv('TRITON_EXTERNAL_CODEGEN_BACKENDS')
+    backend_dirs = os.getenv('TRITON_EXTERNAL_CODEGEN_BACKEND_DIRS')
+
+    if backends is None or backend_dirs is None:
+        return []
+
+    backends = backends.strip().split(';')
+    backend_dirs = backend_dirs.strip().split(';')
+
+    if len(backends) != len(backend_dirs):
+        raise Exception(f"External backend names and directories mismatch")
+
+    ret = []
+
+    for backend, root_dir in zip(backends, backend_dirs):
+        curr_path = os.path.join(root_dir, backend)
+        backend_path = os.path.abspath(os.path.join(curr_path, "backend"))
+        install_dir = os.path.join(os.path.dirname(__file__), "triton", "backends", backend)
+
+        # check conditions
+        assert backend in os.listdir(root_dir), f"{backend} is requested for install but not present in {root_dir}"
```

Comment:
not sure the right philosophy here of assert vs. exception vs. error message., but if we're validating user input assert seems wrong to me.